### ISSUE TITLE
CLOUDSTACK-9341 volume import and export when using swift as secondary storage

### ISSUE THREAD

**Issue by skattoju4** [apache/cloudstack#4874]
##### ISSUE TYPE

 * Bug Report
 * Improvement Request

##### COMPONENT NAME

~~~
Secondary Storage
~~~

##### CLOUDSTACK VERSION

~~~
4.7 and above
~~~

##### CONFIGURATION

CloudStack 4.12, XenServer and Swift as Secondary Storage.

##### OS / ENVIRONMENT

N/A

##### SUMMARY

*Uploading a volume fails when using swift as secondary storage.
The error returned is: org.apache.cloudstack.storage.to.VolumeObjectTO cannot be cast to org.apache.cloudstack.storage.to.TemplateObjectTO
*Exporting or downloading a volume fails when using swift as secondary storage.
Cloudstack shows a url that points to a symlink on the secondary stroage vm however the symlink's path is incorrect path so the volume is not downloaded.

##### STEPS TO REPRODUCE

1. Upload a volume from a url when using swift as secondary storage. This will fail.
2. Download a volume when using swift as secondary storage. This will fail.

~~~

~~~

##### EXPECTED RESULTS

1. Volume is successfully uploaded and eventually gets to ready state.
2. A url is shown which points to the volume to download.
~~~

~~~

##### ACTUAL RESULTS

1. The error returned is: 
~~~
org.apache.cloudstack.storage.to.VolumeObjectTO cannot be cast to org.apache.cloudstack.storage.to.TemplateObjectTO
~~~
2. The url shown points at a broken symlink.

**Comment by DaanHoogland** [comment]
@skattoju3 can you add any stacktraces you see in the logs, and possibly from the cloud.log in the ssvm?

**Comment by skattoju4** [comment]
hey daan, i've added the stack trace from the management server. It fails before getting to the ssvm...

**Comment by DaanHoogland** [comment]
tnx @skattoju3 , looks like a simple case of mistaken identity, the code assumes only templates are being uploaded and never volumes.
Can you verify that uploading a template works, please?

**Comment by skattoju4** [comment]
Yes template upload works, however download progress is not reported...

**Comment by rohityadavcloud** [comment]
cc @shwstppr @pdion891

**Comment by SakshiT25** [comment]
Hi, I'm interested in working on this issue. Kindly guide me for the further steps.

**Comment by DaanHoogland** [comment]
> Hi, I'm interested in working on this issue. Kindly guide me for the further steps.

@SakshiT25 please see my reply in https://github.com/apache/cloudstack/issues/5686#issuecomment-1274565162

**Comment by DaanHoogland** [comment]
@skattoju4 closed #4875 and closing this as stale, please reopen if you are reviving the effort.
